zooniverse/front-end-monorepo
# Workflow assignment filter

_Please request review from `@zooniverse/frontend` team or an individual member of that team._ 

Package: app-project

Closes #2381

For projects that have workflow assignment enabled, we do not want signed out or signed in volunteers to be able to load workflows they haven't been assigned yet. For these types of projects, workflows have a "level" designation in their configuration (Note: this can simply indicate the order of the progression through the workflows, not necessarily difficulty). Signed out volunteers can only access level 1 workflows. Signed in volunteers can access the workflow up to the level of the workflow they are assigned to. For instance, if someone is assigned to a workflow that has been designated "level" 3, then they can access workflows 1-3.

How to test:

Run yarn dev and load up https://local.zooniverse.org:3000/projects/srallen086/workflow-assignment-testing as a signed out user. The anonymous user should only have access to level 1 workflows:

<img width="380" alt="Screen Shot 2021-09-15 at 3 56 05 PM" src="https://user-images.githubusercontent.com/5016731/133508350-b77a4865-a12e-4d07-8106-e95a32802a47.png">

Imagine, though, that URLs with the workflow ids are shared and this person tries to access the other ones. Next try loading the level 2 or level 3 workflows directly:

Level 2: https://local.zooniverse.org:3000/projects/srallen086/workflow-assignment-testing/classify/workflow/2334
Level 3: https://local.zooniverse.org:3000/projects/srallen086/workflow-assignment-testing/classify/workflow/2619

Both of these should not complete the load of the classifier and instead prompt the workflow selection menu:

<img width="1393" alt="Screen Shot 2021-09-15 at 3 57 59 PM" src="https://user-images.githubusercontent.com/5016731/133508608-2b1bfeef-bd11-4e99-8cfa-39f3682eb5b3.png">

Level 1 should still be accessible by direct link as well:
https://local.zooniverse.org:3000/projects/srallen086/workflow-assignment-testing/classify/workflow/2333

For a signed in user, try signing in as zootester1. This user has access to levels 1 and 2:

<img width="375" alt="Screen Shot 2021-09-15 at 3 59 57 PM" src="https://user-images.githubusercontent.com/5016731/133508845-16b9b4ad-e726-4f7a-9436-fa5b02fe1ba5.png">

Trying to load level 3 directly should result in the classifier not fully loading and the prompt of the workflow selection menu:

https://local.zooniverse.org:3000/projects/srallen086/workflow-assignment-testing/classify/workflow/2619

<img width="1440" alt="Screen Shot 2021-09-15 at 4 01 01 PM" src="https://user-images.githubusercontent.com/5016731/133508959-cae86bd8-50d0-4a57-8a86-2b8979c8ba59.png">

Levels 1 and 2 should load as expected.

The last testing piece is that as signed in as zootester1 is to then sign out. If workflows labeled as level 2 or 3 are loaded, then this should result in the workflow selector prompt appearing:

<img width="1351" alt="Screen Shot 2021-09-15 at 4 04 19 PM" src="https://user-images.githubusercontent.com/5016731/133509422-83fe0dcf-1225-4090-9286-1b62ccfd9932.png">


Note: as a side effect of writing tests for this, I had to use Enzyme's mount, which necessitating a lot of mock. I did investigate switching over to RTL for these, but doing this on the `ClassifyPage` ended up meaning not only mocking the project app store, but also the classifier library, which was a deep dive I don't have time for unfortunately. In addition, some side effects include switching a couple of child components to use `useRouter` which is easier to mock than the older `withRouter` HOC.

# Review Checklist

## General

- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?
